v0.12 fork choice update

paulhauner opened the pull request:
## Issue Addressed

- Resolves #1054
- Resolves #1059
- Resolves #1060

## Proposed Changes

- Update fork choice to `v0.12.0`
- Refactors everything above `proto_array` to read more like the spec.
- Moves the `beacon_chain::ForkChoice` struct into a new `consensus/fork_choice` crate (along with a significant refactor).
- Renames the `proto_array_fork_choice` crate to `proto_array` (it seems less confusing now we have a `fork_choice` crate too).
- Removes `VerifiedForkChoiceAttestation` and shifts that logic into the new `fork_choice` crate.
  - Replaces `IntoForkChoiceVerifiedAttestation` with `SignatureVerifiedAttestation`.
- Moves `beacon_chain/tests/attestation_verification.rs` tests related to fork choice into the new `fork_choice` crate
  - Also refactors them into @michaelsproul's delicious builder-y-testing-struct pattern.
- Lifts all concurrency locks from `ProtoArray` and `ForkChoice` up to a single, monolithic lock on `beacon_chain.fork_choice`.
  - This makes it easier to reason about the consistency of fork choice at the risk of blocking some legit concurrent operations.
- Hold a write-lock on `beacon_chain.fork_choice` until an block has been added to fork choice *and* leveldb.
  - Once again, favours consistency at the risk of blocking some legit concurrent ops.
- Adds some new methods to `BeaconChainHarness` (nothing breaking).
- Removes the `beacon_chain::Builder::reduced_tree_fork_choice` method:
  - It was incorrectly named.
  - We *must* call it, so I just moved its functionality into `Builder::build`

## TODO

- [x] Delay consideration of attestations from this slot.
- [x] New verification conditions
- [x] ~~Harmony fork choice tests~~ (deferred to later PR)
- [x] ~~Fix this todo from #1228:~~ https://github.com/sigp/lighthouse/blob/f498081941b0b26e5cec13783cc3f3531cdd3e37/beacon_node/beacon_chain/tests/attestation_verification.rs#L123-L125
- [x] Fix double clone in best justified checkpoint
- [x] Rethink caching justified state balances
- [x] Add metrics again

michaelsproul commented:
Do you mind taking the `TODO`s related to fork choice in attestation verification in as part of this PR?

https://github.com/sigp/lighthouse/blob/cb26ddebb1b9ec0ba24da287a4ed83fc1b4d9153/beacon_node/beacon_chain/src/attestation_verification.rs#L554-L561

https://github.com/sigp/lighthouse/blob/cb26ddebb1b9ec0ba24da287a4ed83fc1b4d9153/beacon_node/beacon_chain/src/attestation_verification.rs#L529-L537

I think it would make most sense for the target verification to be added alongside these fork-choice changes, as the underlying fork choice API seems to be changing a bit (and we're getting `get_ancestor`).

michaelsproul commented:
> Fix this todo from #1228:

I already fixed this in #1236, so I'm ticking it off (it's not fork choice related)

paulhauner commented:
This is ready for review, assuming it passes CI :)

paulhauner commented on consensus/fork_choice/src/fork_choice.rs:
```diff
@@ -0,0 +1,892 @@
+/// ## Notes
+///
+/// This function should only ever be passed a `time` that is less than, equal to or one greater
+/// than the previously passed value. I.e., it must be called each time the slot changes.
```
Yep that's right. Will remove this note.

paulhauner commented on consensus/fork_choice/src/fork_choice.rs:
```diff
@@ -0,0 +1,892 @@
+    /// state must fulfil the following conditions:
+    ///
+    /// 1. It must be from `block_root` or an **descendant** of `block_root`.
+    /// 2. It must have a equal to or greater than `ancestor_slot`.
```
Turns out that section of comment was outdated, so I removed it :)

paulhauner commented on consensus/fork_choice/src/fork_choice.rs:
```diff
@@ -0,0 +1,892 @@
+            });
+        }
+
+        // Attestation target must be for a known block.
```
I added a note :)

paulhauner commented:
All comments addressed! There are two interesting commits since your review:

- https://github.com/sigp/lighthouse/pull/1229/commits/17c8fa51a74c8e55f8708bab48f91a651666104c: Makes some changes to `get_ancestor`:
  - Fixes a bug with skip slots
  - Returns an `Option` so we can gracefully detect when we've iterated outside of proto array.
- https://github.com/sigp/lighthouse/pull/1229/commits/1b6ccb05979461015996f1d7d4cee5d9407f34a3: Addresses comments from @michaelsproul.

michaelsproul commented on beacon_node/beacon_chain/tests/attestation_verification.rs:
```diff
@@ -87,7 +87,7 @@ fn get_valid_unaggregated_attestation<T: BeaconChainTypes>(
 fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
     chain: &BeaconChain<T>,
-    aggregate: Attestation<T::EthSpec>,
+    mut aggregate: Attestation<T::EthSpec>,
```
These were part of the temporary FIXME change, and will trigger warnings now, so should be reverted (I like `git checkout -p` for this)

michaelsproul commented on beacon_node/beacon_chain/tests/attestation_verification.rs:
```diff
@@ -97,10 +97,11 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
     let committee_len = committee.committee.len();
 
-    let (aggregator_index, aggregator_sk) = committee
+    let (aggregator_committee_pos, aggregator_index, aggregator_sk) = committee
```
Can delete

michaelsproul commented on beacon_node/beacon_chain/tests/attestation_verification.rs:
```diff
@@ -97,10 +97,11 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
         .committee
         .iter()
-        .find_map(|&val_index| {
+        .enumerate()
+        .find_map(|(committee_pos, &val_index)| {
```
Can revert

michaelsproul commented on beacon_node/beacon_chain/tests/attestation_verification.rs:
```diff
@@ -112,7 +113,7 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
 
             if proof.is_aggregator(committee_len, &chain.spec).unwrap() {
-                Some((val_index, aggregator_sk))
+                Some((committee_pos, val_index, aggregator_sk))
```
Not reqd

paulhauner commented:
This one should revert that old FIXME change :)

https://github.com/sigp/lighthouse/pull/1229/commits/00bb20a42378f7a3ef71ad9d166e020e3ec02c43
